piker/brokers/kraken.py

```diff
@@ -275,14 +438,321 @@ async def bars(
+            if float(vols[ticker]) != 0:
+                msg = pack_position(acc_name, norm_sym, pos, vols[ticker])
+                all_positions.append(msg.dict())
+
+        ## TODO: create a new ems message schema for open orders
+        open_orders = await client.kraken_endpoint('OpenOrders', {})
+        print(open_orders)
+        #await tractor.breakpoint()
+
+        await ctx.started((all_positions, (acc_name,)))
+
+        # Get websocket token for authenticated data stream
+        # Assert that a token was actually received 
+        resp = await client.kraken_endpoint('GetWebSocketsToken', {})
+        assert resp['error'] == []
+        token = resp['result']['token']
+
+        async with (
+            ctx.open_stream() as ems_stream,
+            trio.open_nursery() as n,
+        ):
+            ## TODO: maybe add multiple accounts
+            n.start_soon(handle_order_requests, client, ems_stream)
+
+            # Process trades msg stream of ws
+            async with open_autorecon_ws(
+                'wss://ws-auth.kraken.com/',
+                fixture=subscribe,
+                token=token,
+            ) as ws:
+                from pprint import pprint
+                async for msg in process_trade_msgs(ws):
+                    pprint(msg)
+                    for trade in msg:
+                        # check the type of packaged message
+                        assert type(trade) == Trade
+                        # prepare and send a status update for line update
+                        trade_msg = BrokerdStatus(
+                            reqid=trade.reqid,
+                            time_ns=time.time_ns(),
+
+                            account='kraken.spot',
+                            status='executed',
+                            filled=float(trade.size),
+                            reason='Order filled by kraken',
+                            # remaining='' ## TODO: not sure what to do here.
+                            broker_details={
+                                'name': 'kraken',
+                                'broker_time': trade.broker_time
+                            }
+                        )
+                        
+                        await ems_stream.send(trade_msg.dict())
+
+                        # send a fill msg for gui update
+                        fill_msg = BrokerdFill(
+                            reqid=trade.reqid,
+                            time_ns=time.time_ns(),
+
+                            action=trade.action,
+                            size=float(trade.size),
+                            price=float(trade.price),
+                            ## TODO: maybe capture more msg data i.e fees?
+                            broker_details={'name': 'kraken'},
+                            broker_time=float(trade.broker_time)
+                        )
+                        
+                        await ems_stream.send(fill_msg.dict())
```

Comment:
> so I had to add this to get the triangles to show. I tried its on its own, nothing (line remained, but no triangle), I tried just the status update (line gone, but no triangle), and then after digging into the code, I decided to try both, and that worked.

Ahh yeah could be. If that's the case then all good.

> fyi, I noticed in your comments you mentioned potentially getting rid of the BrokerdFill msg as it seems redundant from ib, this would break the triangles function of the ui as it needs both msgs to work.

Yeah then let's keep it for now. I think the `ib` case is weird as well because they send multiple messages for all sorts of stuff. For example there's like a separate msg for the commissions report. I think it also makes sense to have fill msgs be distinct since you can definitely have multiple fills *complete* and order execution depending on book liquidity.